10-JavaScript/classes.md

```diff
@@ -0,0 +1,289 @@
+
+```javascript
+class Book {
+  constructor(title, author) {
+    this.title = title;
+    this.author = author;
+  }
+
+  toString() {
+   return `${this.title} by ${this.author}`;
+  }
+}
+
+let warAndPeace = new Book("War and Peace", "Leo Tolstoy");
+
+console.log(warAndPeace.toString());
+> War and Peace by Leo Tolstoy
+```
+
+Classes can also be declared as an expression, similar to a function.
+
+```javascript
+const Book = class {
+ // code
+};
+```
+
+
+## Methods
+
+In a JavaScript class, methods are functions declared inside the class, without the `function` keyword.  There are a number of special functions such as __constructor__, __getter methods__ and __setter methods__.
+
+### Constructor Methods
+
+The constructor method is a special method named `constructor` which is called when a new instance is created.  You can think of it like an `initialize` method from Ruby.  Constructors exist to start our instances off in a proper state.  Use them to set your attributes with initial values using `this.<attributeName>` to create attributes (similar to using @ in Ruby).
+
+```javascript
+class Book {
+  constructor(title, author) {
+    console.log("Constructor Ran!");
+    this.title = title;
+    this.author = author;
+  }
+}
+
+let poodr = new Book("Practical Object Oriented Programming in Ruby", "Sandy Mietz");
+
+> Constructor Ran!
+```
+
+### Instance Methods
+
+Instance methods are created like a constructors with their own names.  
+
+```javascript
+const SALESTAX = 0.08;
+
+class Book {
+  constructor(title, author, price) {
+    this.title = title;
+    this.author = author;
+    this.price = price;
+  }
+  totalPrice() {
+    return this.price * (1.0 + SALESTAX);
+  }
+}
+
+var adaBook = new Book("Poodr", "Sandy Meitz", 100);
+console.log(adaBook.totalPrice());
+> 108
+```
+
+### Getter & Setter Methods
+
+Getter methods are methods which are run when a property with the same name is accessed.  Similarly a Setter method is called when there is an attempt to change a property.  
+
+```javascript
+class Book {
+  constructor(title, author) {
+    console.log("Constructor Run!");
+    this._title = title;
+    this._author = author;
+  }
+
+  // getter methods
+  get author() {
+    return this._author;
+  }
+
+  // setter method
+  set author(newAuthor) {
+    if (typeof newAuthor === "string") {
+      this._author = newAuthor;
+    }
+  }
+
+}
+let poodr = new Book("Practical Object Oriented Programming in Ruby", "Sandy Meitz");
+
+// setter method
+poodr.author = "Chris";
+
+console.log(`${poodr.author}`); // getter method
+> Chris
+```
+
+Notice that the properties are all named __\_propertyName__.  This is a typical naming convention for instance variables you want to keep private.  It's a convention that signals an intent to prevent access outside the class as nothing is really private in JavaScript.  Properties cannot have the same names as our getter and setter methods as doing so will result in infinite recursion.  
```

Comment:
I think it would make sense to have this note before the code - it would be less confusing in parsing this example IMO.